index.html

```diff
@@ -0,0 +1,451 @@
+  query: Query
+}
+
+enum csdl_Graph {
+  ROCKETS    @csdl_endpoint(url: "https://rockets.example.com"),
+  ASTRONAUTS @csdl_endpoint(url: "https://astronauts.example.com"),
+}
+
+type Query {
+  rockets: [Rocket]!       @csdl_resolve(graph: ROCKETS)
+  astronauts: [Astronaut]! @csdl_resolve(graph: ASTRONAUTS)
+}
+
+type Astronaut
+  @csdl_key(graph: ASTRONAUTS, repr: "{ id }")
+  @csdl_key(graph: ROCKETS, repr: "{ id }")
+{
+  id: String!     @csdl_resolve(graph: ASTRONAUTS)
+  name: String!   @csdl_resolve(graph: ASTRONAUTS)
+  tripId: String! @csdl_resolve(graph: ASTRONAUTS)
+  rocket: Rocket! @csdl_resolve(graph: ASTRONAUTS)
+}
+
+type Rocket 
+  @csdl_key(graph: ASTRONAUTS, repr: "{ id }")
+  @csdl_key(graph: ROCKETS, repr: "{ id }")
+{
+  id: String!           @csdl_resolve(graph: ROCKETS)
+  name: String!         @csdl_resolve(graph: ROCKETS)
+  astronaut: Astronaut! @csdl_resolve(graph: ASTRONAUTS)
+  captain: Astronaut    @csdl_resolve(graph: ASTRONAUTS,
+                                      provides: "{ tripId }")
+}
+```
+
+The meaning of these directives is explored the [Directives](#directives) section.
+
+## Actors
+
+```mermaid
+#::[Actors and roles within an example composition pipeline]
+flowchart TB
+    subgraph A [subgraph A]
+      schemaA([schema A])
+      endpointA([endpoint A])
+    end    
+    subgraph B [subgraph B]
+      schemaB([schema B])
+      endpointB([endpoint B])
+    end
+    subgraph C [subgraph C]
+      schemaC([schema C])
+      endpointC([endpoint C])
+    end
+    subgraph Producer
+      Composer
+    end
+    CSDL([CSDL])
+    subgraph Consumer
+      Router
+    end    
+    A-->Composer
+    B-->Composer
+    C-->Composer
+    Composer-->CSDL
+    CSDL-->Router
+    Router-->published([Published Schema])
+    published-->Clients
+    Clients-->Router
+```
+
+- **Producers** generate CSDL. This spec places requirements on csdl producers.
+- **Consumers** consume CSDL. This spec places requirements on csdl consumers.
+- **Composers** or **compilers** are producers which compose subgraph schemas into CSDL. This document places no particular requirements on the composition algorithm, except that it must produce valid csdl.
+- **Routers** are consumers serve CSDL as a GraphQL endpoint. Graph routers differ from standard GraphQL endpoints in that they are not expected to process data or communicate with (non-GraphQL) backend services on their own. Instead, graph routers receive GraphQL requests and service them by performing additional GraphQL requests. This spec provides guidance for implementing routers, but does not require particular implementations of query separation or dispatch.
+
+Other components of the system include:
+- **Endpoints** are running servers which can resolve GraphQL queries against a schema. In this version of the spec, endpoints must be URLs, typically http/https URLs.
+- **Subgraphs** are the GraphQL schemas which were composed to form the CSDL. A subgraph has:
```

Comment:
Should it be documented that subgraph schemas are *not* expected to be CSDL, and in fact it (may be? is?) an error for a subgraph schema to contain csdl_Graph?

Reply:
oh, hmm! i *don't* think it should be an error for those documents to provide a composition layer. If you feed a schema with CS directives as input into a composer which doesn't expect them, those directives just shouldn't matter at all—you'll get back a composed schema which routes through another composed schema, which is maybe non-ideal, but is functionally fine. In practice, I think we have an opportunity to make the composer CS-aware for input, so if you do compose already-composed schema, the output schema is able to "flatten" the graph. With a CS-aware composer, if `A` and `B` both compose `C` and we compose `A` and `B`, we get back a graph which composes `A`, `B`, `C`, rather than one that composes just `A` and `B`. We do not have to implement this, but I don't think we should preclude it—I could see it being quite valuable for larger orgs who would like to have multiple layers of composition.

Reply:
I agree that it shouldn't be an error for subgraphs to present CSDL to a composer, although I'd expect that since the composer is generating CSDL as output, that it would remove any of the CSDL directives and types present in the subgraphs. The flattening idea sounds interesting, but it requires a decent amount of cooperation from the composer, graph router, and computing environment, and I'd imagine it's outside the scope of the spec.

Reply:
yeah, all composer behavior is outside the scope of this spec, i just don't want to foreclose it